### PR TITLE
create a basic design patterns document

### LINKED ISSUE #11: Write some best practices guidance for creating components
Within `vdom`, the best way (that I can think of right now) to create new elements is to create functions that return elements:

```python
from vdom import h, p, div
def happiness(level=90):
  smiley = "😃"
  percent = level / 100.
    
  if(percent < 0.25):
    smiley = "☹️"
  elif(percent < 0.50):
    smiley = "😐"
  elif(percent < 0.75):
    smiley = "😀"

    
  return p([
      'Happiness ', smiley, h('br'), h('meter', level, min="0", low=25, high=75, optimum=90, max=100, value=level)
  ])
# Usage:

happiness()

div([
  happiness(10),
  happiness(32),
  happiness(65),
  happiness(80)
])
```

Example in nteract:

screen shot 2017-09-25 at 11 49 15 am

### PR DESCRIPTION
**Pull request by rgbkrk**
Closes #11, at least in a basic way for now.

### CODE REVIEW

File: docs/design-patterns.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
+# Design Patterns
+
+The main operating principle for `vdom` is:
+
+> Write functions that return `vdom` elements
+
+In matching with `React` parlance, we'll call these functions components. This allows you to share, remix, and use components amongst everyone. You'll be able to compose together components to create greater layouts than you have
```
Comment (willingc) [inline docs/design-patterns.md]: **components**

File: docs/design-patterns.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
+# Design Patterns
+
+The main operating principle for `vdom` is:
```
Comment (willingc) [inline docs/design-patterns.md]: `vdom` (`virtual DOM`)

The non-frontend folks will thank you for the definition.

File: docs/design-patterns.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+> Write functions that return `vdom` elements
+
+In matching with `React` parlance, we'll call these functions components. This allows you to share, remix, and use components amongst everyone. You'll be able to compose together components to create greater layouts than you have
+before.
+
+## Introductory component
+
+We'll start with a component that takes a level of happiness that produces a light visualization:
```
Comment (willingc) [inline docs/design-patterns.md]: ...happiness ~~that~~ and

File: docs/design-patterns.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    smiley = "😀"
+
+    
+  return span(
+      p('Happiness ', smiley),
+      meter(level, min=0, low=25, high=75, optimum=90, max=100, value=level)
+  )
+```
+
+The user of the component only has to know how to call `happiness` with their level of happiness:
```
Comment (willingc) [inline docs/design-patterns.md]: The user of the `happiness` component only needs to call it with a level of happiness from 0 to 100.

File: docs/design-patterns.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+<span>
+<p>Happiness 😃</p>
+<meter min="0" low="25" high="75" optimum="90" max="100" value="96">
+  96
+</meter>
+</span>
+
+------------
+
+:tada: Our first component is ready! Since we can think of these as little building block components, we can put several of these together to create whole layouts:
```
Comment (willingc) [inline docs/design-patterns.md]: ~~these~~ `happiness` as a little building ...

File: docs/design-patterns.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+<p>Happiness 😃</p>
+<meter min="0" low="25" high="75" optimum="90" max="100" value="80">
+  80
+</meter>
+</span>
+
+-------------------
+
+
+## Working with python objects
```
Comment (willingc) [inline docs/design-patterns.md]: Python

File: docs/design-patterns.md [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+For this section, you'll need `ggplot` and `matplotlib`. We'll create a component
```
Comment (willingc) [inline docs/design-patterns.md]: `matplotlib` packages installed. We'll create a component, `fancy_hist`, that ... side:

File: docs/design-patterns.md (hunk 2 above)
Comment (rgbkrk) [reply docs/design-patterns.md]: Oh right! Thank you.

**Comment by rgbkrk** [comment]
Great, addressed feedback. Thanks for the suggestions!

**Comment by willingc** [comment]
Thanks @rgbkrk

File: docs/design-patterns.md (hunk 2 above)
Comment (rgbkrk) [reply docs/design-patterns.md]: Reflecting on this today I wish I called the library `layout` or something. I guess someone could build on top of `vdom` and abstract away the innards here.

**Comment by mpacer** [comment]
Is this open in theory to further edits? I feel like if we are really making this analogous with react it would be good to pass in only keyword arguments in your functions for those things being passed through that aren't supposed to be interpreted directly and literally. That way it's closer to children vs. general props.

Also I can't think of a way to analogise to state. But I think it would be valuable to think about how this could be brought to include classes as well. In that case I could imagine maybe generalising the notion of state managment by the use of a class level context manager. Maybe then, the notion of enforcing not mutating your props but allowing state mutation could be addressed by passing around function handles from instance parents to children. I think the challenge there is whether python has a way to express the noninheritance version of the parent child relationship with instance creation inside

**Comment by rgbkrk** [comment]
This is definitely open to edits. It was my first take at it. I feel like the keyword argument standard is a good thing on first glance though I feel like we don't want to explicitly prevent people for their custom functions. You're making me think about how custom functions could work with children though 🤔

**Comment by rgbkrk** [comment]
As for state I'm a tad worried we don't have a parallel because we're not actually managing a virtual dom (only in name) and we don't push updates for changes. I feel like there's more than one way to handle what a magic updating one would do. 🤔

Definitely all good things to explore for the doc

**Pull request merged**